### PR TITLE
AppInsightTelemetry support for CNI

### PR DESCRIPTION
**Pull request by tamilmani1989**
**What this PR does / why we need it**:
This PR adds AI telemetry support for CNI via telemetry service. The CNI will send telemetry data to telemetry service and telemetry service in turn will send data to AppInsight.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: telemetry/aiwrapper.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+import (
+	"github.com/Azure/azure-container-networking/aitelemetry"
+	"github.com/Azure/azure-container-networking/log"
+)
+
+var (
+	aiMetadata     string
+	th             aitelemetry.TelemetryHandle
+	gDisableAll    bool
```
Comment (jaer-tsun) [inline telemetry/aiwrapper.go]: what's the g prefix for?

File: telemetry/aiwrapper.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	}
+
+	if cnireport.Metric.Name != "" && !gDisableMetric {
+		th.TrackMetric(cnireport.Metric)
+	} else if !gDisableTrace {
+		sendReport(cnireport)
+	}
+}
+
+func CloseAITelemetryHandle() {
```
Comment (jaer-tsun) [inline telemetry/aiwrapper.go]: check if th is nil

File: cni/network/network.go [hunk 3]
```diff
@@ -186,6 +190,23 @@ func (plugin *netPlugin) getPodInfo(args string) (string, string, error) {
 	return k8sPodName, k8sNamespace, nil
 }
 
+func setCustomDimensions(cniMetric *telemetry.CNIReport, nwCfg *cni.NetworkConfig, err error) {
```
Comment (jaer-tsun) [inline cni/network/network.go]: not sure if you need to check if the other pointers are nil here

File: telemetry/aiwrapper.go (hunk 1 above)
Comment (tamilmani1989) [reply telemetry/aiwrapper.go]: g for global..already using disableAll in CreateAITelemetryHandle..added this to prevent conflict

File: cni/network/network.go (hunk 3 above)
Comment (tamilmani1989) [reply cni/network/network.go]: no worries..not needed..nothing can be nil..

File: cni/network/network.go [hunk 4]
```diff
@@ -276,6 +307,8 @@ func (plugin *netPlugin) Add(args *cniSkel.CmdArgs) error {
 		return err
 	}
 
+	plugin.report.ContainerName = k8sPodName + ":" + k8sNamespace
```
Comment (ashvindeodhar) [inline cni/network/network.go]: what's this line doing / needed?

File: cni/network/plugin/main.go [hunk 5]
```diff
@@ -214,6 +215,8 @@ func main() {
 	tb.ConnectToTelemetryService(telemetryNumRetries, telemetryWaitTimeInMilliseconds)
 	defer tb.Close()
 
+	netPlugin.SetCNIReport(cniReport, tb)
```
Comment (ashvindeodhar) [inline cni/network/plugin/main.go]: where is tb declared?

File: telemetry/telemetrybuffer.go [hunk 6]
```diff
@@ -28,6 +28,14 @@ import (
 // TelemetryConfig - telemetry config read by telemetry service
 type TelemetryConfig struct {
 	ReportToHostIntervalInSeconds time.Duration `json:"reportToHostIntervalInSeconds"`
+	DisableAll                    bool
```
Comment (ashvindeodhar) [inline telemetry/telemetrybuffer.go]: where is this getting used?

File: telemetry/telemetry.go [hunk 7]
```diff
@@ -383,3 +385,21 @@ func (reportMgr *ReportManager) ReportToBytes() ([]byte, error) {
 	report, err = json.Marshal(reportMgr.Report)
 	return report, err
 }
+
+func SendCNIMetric(cniReport *CNIReport, tb *TelemetryBuffer) error {
```
Comment (ashvindeodhar) [inline telemetry/telemetry.go]: instead of creating different functions for CNS, such as SendCNSMetric, can this be just one function like SendMetric and takes the type / *Report.
Can it take *Report to reuse the func signature between CNI, CNS ?

File: telemetry/aiwrapper.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
+package telemetry
```
Comment (ashvindeodhar) [inline telemetry/aiwrapper.go]: need copyright?

File: telemetry/aiwrapper.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+var (
+	aiMetadata     string
+	th             aitelemetry.TelemetryHandle
+	gDisableTrace  bool
+	gDisableMetric bool
+)
+
+const (
+	waitTimeInSecs = 10
```
Comment (ashvindeodhar) [inline telemetry/aiwrapper.go]: can you add a comment

File: telemetry/aiwrapper.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+		log.Printf("Telemetry is disabled")
+		return
+	}
+
+	th = aitelemetry.NewAITelemetry(aiMetadata, aiConfig)
+	gDisableMetric = disableMetric
+	gDisableTrace = disableTrace
+}
+
+func sendReport(cnireport CNIReport) {
```
Comment (ashvindeodhar) [inline telemetry/aiwrapper.go]: are we going to create different function for CNS?

File: telemetry/aiwrapper.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+	report.CustomDimensions[ContextStr] = cnireport.Context
+	report.CustomDimensions[SubContextStr] = cnireport.SubContext
+	report.CustomDimensions[VMUptimeStr] = cnireport.VMUptime
+	report.CustomDimensions[OperationTypeStr] = cnireport.OperationType
+	report.CustomDimensions[VersionStr] = cnireport.Version
+
+	th.TrackLog(report)
+}
+
+func SendAITelemetry(cnireport CNIReport) {
```
Comment (ashvindeodhar) [inline telemetry/aiwrapper.go]: are we going to create different function for CNS?

File: telemetry/azure-vnet-telemetry.config [hunk 12]
```diff
@@ -1,3 +1,11 @@
...
+    "reportToHostIntervalInSeconds": 30,
+    "BatchSizeInBytes":16384,
+    "BatchIntervalInSecs":15,
+    "RefreshTimeoutInSecs": 15,
+    "DisableMetadataThread": false,
+    "DisableAll": false,
+    "DisableTrace": false,
+    "DisableMetric": false,
+    "DebugMode":true
 }
```
Comment (ashvindeodhar) [inline telemetry/azure-vnet-telemetry.config]: Need a new line at the end?

File: telemetry/telemetry.go [hunk 13]
```diff
@@ -383,3 +385,21 @@ func (reportMgr *ReportManager) ReportToBytes() ([]byte, error) {
...
+	var err error
```
Comment (ashvindeodhar) [inline telemetry/telemetry.go]: Can you change to
var (
...
...
)

File: telemetry/constants.go [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+	CNIUpdateTimeMetricStr = "CNIUpdateTimeMs"
+
+	// Dimension Names
+	ContextStr        = "Context"
+	SubContextStr     = "SubContext"
+	VMUptimeStr       = "VMUptime"
+	OperationTypeStr  = "OperationType"
+	VersionStr        = "Version"
+	StatusStr         = "Status"
+	CNIModeStr        = "CNIMode"
```
Comment (ashvindeodhar) [inline telemetry/constants.go]: what is CNI mode?

File: telemetry/constants.go [hunk 15]
```diff
@@ -0,0 +1,27 @@
+package telemetry
```
Comment (ashvindeodhar) [inline telemetry/constants.go]: copyright? Can you check this for all the new files you are adding for AI related PRs?

File: telemetry/aiwrapper.go [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+import (
+	"github.com/Azure/azure-container-networking/aitelemetry"
+	"github.com/Azure/azure-container-networking/log"
+)
+
+var (
+	aiMetadata     string
+	th             aitelemetry.TelemetryHandle
+	gDisableTrace  bool
```
Comment (ashvindeodhar) [inline telemetry/aiwrapper.go]: what's g style? :)

File: cni/network/network.go (hunk 4 above)
Comment (tamilmani1989) [reply cni/network/network.go]: yes this is for filtering logs based on container name

File: telemetry/telemetrybuffer.go (hunk 6 above)
Comment (tamilmani1989) [reply telemetry/telemetrybuffer.go]: used by telemetry service

File: telemetry/telemetry.go (hunk 7 above)
Comment (tamilmani1989) [reply telemetry/telemetry.go]: nope. CNI uses telemetry service but not cns

File: telemetry/constants.go (hunk 14 above)
Comment (tamilmani1989) [reply telemetry/constants.go]: CNI mode tells single tenancy or multitenancy

File: telemetry/aiwrapper.go (hunk 16 above)
Comment (tamilmani1989) [reply telemetry/aiwrapper.go]: ya replied for jaeryn comment. g for global . im using disableTrace as functional parameter. Inorder to prevent conflict i added g at beginning. let me know if there is better way

File: cni/network/plugin/main.go (hunk 5 above)
Comment (tamilmani1989) [reply cni/network/plugin/main.go]: tb := telemetry.NewTelemetryBuffer("")

File: aitelemetry/telemetrywrapper.go [hunk 17]
```diff
@@ -90,6 +92,14 @@ func NewAITelemetry(
 	id string,
 	aiConfig AIConfig,
 ) TelemetryHandle {
+	env, _ := os.LookupEnv("AZACN_TESTENV")
```
Comment (ashvindeodhar) [inline aitelemetry/telemetrywrapper.go]: why does it say TESTENV ?

File: aitelemetry/telemetrywrapper.go [hunk 18]
```diff
@@ -90,6 +92,14 @@ func NewAITelemetry(
...
+	if env != "test" {
+		cloudName, err := common.GetAzureCloud()
+		if cloudName != azurePublicCloud {
+			debuglog("[AppInsights] This is not azure public cloud.%s err:%v", cloudName, err)
```
Comment (ashvindeodhar) [inline aitelemetry/telemetrywrapper.go]: debuglog - is this a std. convention? I have seen camel case used in general ( i.e. similar to debugLog )

File: aitelemetry/telemetrywrapper.go (hunk 17 above)
Comment (ashvindeodhar) [reply aitelemetry/telemetrywrapper.go]: is it supposed to say AZACN_CLOUDENV ? or just ENV ?

File: aitelemetry/telemetrywrapper.go (hunk 18 above)
Comment (ashvindeodhar) [reply aitelemetry/telemetrywrapper.go]: remove . after cloud = cloud.%s

File: common/utils.go [hunk 19]
```diff
@@ -14,15 +14,17 @@ import (
...
 
 	"github.com/Azure/azure-container-networking/log"
 )
 
 const (
 	metadataURL           = "http://169.254.169.254/metadata/instance?api-version=2017-08-01&format=json"
-	httpConnectionTimeout = 10
-	headerTimeout         = 20
+	azCloudUrl            = "http://169.254.169.254/metadata/instance/compute/azEnvironment?api-version=2018-10-01&format=text"
+	httpConnectionTimeout = 7
```
Comment (ashvindeodhar) [inline common/utils.go]: why changed to 7?

File: common/utils.go [hunk 20]
```diff
@@ -288,3 +290,31 @@ func SaveHostMetadata(metadata Metadata, fileName string) error {
 
 	return err
 }
+
+func GetAzureCloud() (string, error) {
+	req, err := http.NewRequest("GET", azCloudUrl, nil)
```
Comment (ashvindeodhar) [inline common/utils.go]: can you add a defer function that logs the error? If for any reason this function returns an error ( for public cloud ), we would disable the telemetry and that would be bad. It would be good to log and know if this functions errors out for any reason, ideally we would add the metric on this error so that we get ICM and look at the error and correct it so that we start getting the logs correctly.

File: aitelemetry/telemetrywrapper.go (hunk 17 above)
Comment (tamilmani1989) [reply aitelemetry/telemetrywrapper.go]: this is for uni test..just for making sure not conflicting with any other env variable

File: common/utils.go (hunk 19 above)
Comment (tamilmani1989) [reply common/utils.go]: i thought 10 is too long for a timeout. what say

File: common/utils.go (hunk 20 above)
Comment (tamilmani1989) [reply common/utils.go]: im logging the error at calling place..we cannot log metric as this is base condition for sending telemetry

Comment (tamilmani1989) [reply common/utils.go]: one thing we can do is that return error and the calling side can retry.

File: cni/telemetry/service/telemetrymain.go [hunk 21]
```diff
@@ -97,6 +97,14 @@ func setTelemetryDefaults(config *telemetry.TelemetryConfig) {
 	if config.BatchSizeInBytes == 0 {
 		config.BatchSizeInBytes = defaultBatchSizeInBytes
 	}
+
+	if config.GetEnvRetryCount == 0 {
+		config.GetEnvRetryCount = 2
+	}
+
+	if config.GetEnvRetryWaitTimeInSecs == 0 {
+		config.GetEnvRetryWaitTimeInSecs = 3
```
Comment (ashvindeodhar) [inline cni/telemetry/service/telemetrymain.go]: change const ( 2 and 3 ) to something similar to what you have above? e.g. - defaultBatchSizeInBytes

File: cni/network/network.go [hunk 22]
```diff
@@ -244,7 +244,7 @@ func (plugin *netPlugin) Add(args *cniSkel.CmdArgs) error {
 		enableInfraVnet  bool
 		enableSnatForDns bool
 		nwDNSInfo        network.DNSInfo
-		cniMetric        telemetry.CNIReport
+		cniMetric        telemetry.AIMetric
```
Comment (ashvindeodhar) [inline cni/network/network.go]: why was this changed?

File: telemetry/telemetrybuffer.go [hunk 23]
```diff
@@ -60,7 +61,10 @@ const (
 	cni                = "CNI"
 )
 
-var payloadSize uint16 = 0
+var (
+	payloadSize                uint16 = 0
```
Comment (ashvindeodhar) [inline telemetry/telemetrybuffer.go]: I don;t think you need to set this to 0. GO inits variables ( like bool below ). You can choose to keep it / remove it.

File: aitelemetry/telemetrywrapper.go (hunk 17 above)
Comment (ashvindeodhar) [reply aitelemetry/telemetrywrapper.go]: No what I mean is why does the variable name says TESTENV. I thought it's a test environment - which is not the case here. I think it can be a bit misleading

Comment (tamilmani1989) [reply aitelemetry/telemetrywrapper.go]: this is for unit test to pass...unit test can't make call to metadata server..let me check if i can do with mockagent by passing url

File: cni/network/network.go (hunk 22 above)
Comment (tamilmani1989) [reply cni/network/network.go]: i dont want to change existing CNIReport structure as this schema used by hostnetagent

File: telemetry/telemetrybuffer.go (hunk 23 above)
Comment (tamilmani1989) [reply telemetry/telemetrybuffer.go]: i didnt add this..i just moved this
